src/rdkafka_cgrp.c

```diff
@@ -2666,21 +2695,76 @@ static rd_kafka_op_res_t rd_kafka_cgrp_consumer_handle_next_assignment(
         return RD_KAFKA_OP_RES_HANDLED;
 }
 
+static rd_kafka_topic_partition_list_t *
+rd_kafka_cgrp_consumer_assignment_with_metadata(
+    rd_kafka_cgrp_t *rkcg,
+    rd_kafka_topic_partition_list_t *assignment,
+    rd_list_t **missing_topic_ids) {
+        int i;
+        rd_kafka_t *rk = rkcg->rkcg_rk;
+        rd_kafka_topic_partition_list_t *assignment_with_metadata =
+            rd_kafka_topic_partition_list_new(assignment->cnt);
+        for (i = 0; i < assignment->cnt; i++) {
+                struct rd_kafka_metadata_cache_entry *rkmce;
+                rd_kafka_topic_partition_t *rktpar;
+                rd_kafka_Uuid_t request_topic_id =
+                    rd_kafka_topic_partition_get_topic_id(
+                        &assignment->elems[i]);
+                rd_kafka_rdlock(rk);
+                rkmce =
+                    rd_kafka_metadata_cache_find_by_id(rk, request_topic_id, 1);
+
+                if (rkmce) {
+                        rd_kafka_topic_partition_list_add_with_topic_name_and_id(
+                            assignment_with_metadata, request_topic_id,
+                            rkmce->rkmce_mtopic.topic,
+                            assignment->elems[i].partition);
+                        rd_kafka_rdunlock(rk);
+                        continue;
+                }
+                rd_kafka_rdunlock(rk);
+
+                rktpar = rd_kafka_topic_partition_list_find_topic_by_id(
+                    rkcg->rkcg_current_assignment, request_topic_id);
+                if (rktpar) {
+                        rd_kafka_topic_partition_list_add_with_topic_name_and_id(
+                            assignment_with_metadata, request_topic_id,
+                            rktpar->topic, assignment->elems[i].partition);
+                        continue;
+                }
```

Comment:
We need to ack the assignment even if we don't have metadata, and the protocol ensures it will be at least in current assignment. If we remove and don't ack it, the coordinator could discard the ack.

Later the partitions for that topic will be removed by the coordinator